datashader/datatypes.py

```diff
@@ -0,0 +1,364 @@
+import numpy as np
+from pandas.api.extensions import (
+    ExtensionDtype, ExtensionArray, register_extension_dtype)
+from numbers import Integral
+
+
+@register_extension_dtype
+class RaggedDtype(ExtensionDtype):
+    name = 'ragged'
+    type = np.ndarray
+    base = np.dtype('O')
+
+    @classmethod
+    def construct_array_type(cls):
+        return RaggedArray
+
+    @classmethod
+    def construct_from_string(cls, string):
+        if string == 'ragged':
+            return RaggedDtype()
+        else:
+            raise TypeError("Cannot construct a '{}' from '{}'"
+                            .format(cls, string))
+
+
+class RaggedArray(ExtensionArray):
+    def __init__(self, data, dtype=None):
+        """
+        Construct a RaggedArray
+
+        Parameters
+        ----------
+        data
+            List or numpy array of lists or numpy arrays
+        dtype: np.dtype or str or None (default None)
+            Datatype to use to store underlying values from data.
+            If none (the default) then dtype will be determined using the
+            numpy.result_type function
+        """
+        if (isinstance(data, dict) and
+                all(k in data for k in
+                    ['mask', 'start_indices', 'flat_array'])):
+
+            self._mask = data['mask']
+            self._start_indices = data['start_indices']
+            self._flat_array = data['flat_array']
+        else:
+            # Compute lengths
+            index_len = len(data)
+            buffer_len = sum(len(datum)
+                             if datum is not None
+                             else 0 for datum in data)
+
+            # Compute necessary precision of start_indices array
+            for nbits in [8, 16, 32, 64]:
+                start_indices_dtype = 'uint' + str(nbits)
+                max_supported = np.iinfo(start_indices_dtype).max
+                if buffer_len <= max_supported:
+                    break
+
+            # infer dtype if not provided
+            if dtype is None:
+                dtype = np.result_type(*[np.atleast_1d(v)
+                                         for v in data
+                                         if v is not None])
+
+            # Initialize representation arrays
+            self._mask = np.zeros(index_len, dtype='bool')
+            self._start_indices = np.zeros(index_len, dtype=start_indices_dtype)
+            self._flat_array = np.zeros(buffer_len, dtype=dtype)
+
+            # Populate arrays
+            next_start_ind = 0
+            for i, array_el in enumerate(data):
```

Comment:
Should this constructor be numba-ized?

Reply:
At this point I expect that `data` will be a list of lists or a numpy array of numpy arrays (with `dtype='object'`), so I didn't expect numba to be able to do much with it (at least I didn't think I could use numba in `nopython` mode).

Reply:
Ok.